### DISCUSSION TITLE
How to get model predict and build graphs

### DISCUSSION THREAD

**Discussion by AlineGiroux** [jmsigner/amt#67] [Q&A] [answered]
I am using amt to run issf analysis and I am not sure how can I get the predict of the model to build my graphs. 
I am sedding location data (resu.txt), map (PANTANAL-2015-MAPBIOMAS) and code below. 
I pretend to creat a graph of selection strength (similar to plot 3 - ssf.jpg), and I also pretend to plot the histograms of step lengths and turn angles. 
Can you help me please? @bsmity13 @jmsigner 

[resu.txt](https://github.com/jmsigner/amt/files/8565869/resu.txt)
[PANTANAL-2015-MAPBIOMAS.zip](https://github.com/jmsigner/amt/files/8565909/PANTANAL-2015-MAPBIOMAS.zip)

```
##############################################################################
################PART 1 - GENERATE RANDOM STEPS USING AMT######################
#Load data
resu = read.table("resu.txt",sep=",",h=T)
resu$date = as.POSIXct(resu$date,format="%Y-%m-%d %H:%M:%S",tz="Etc/GMT-4")
head(resu)

#Load package 
library(amt)

#1.1) PREPARE MOVEMENT DATA 
#create a track_xyt object (amt object)
tra=track(resu$x,resu$y,resu$date,id_=resu$ID,tpt=resu$tpt,hour=resu$hour, states=resu$states,order_by_ts=T,check_duplicates = F,crs=CRS("+proj=utm +zone=21 +south"))
head(tra)
#check sampling rate
sampling.rate  <- list()
for(i in 1:length(unique(tra$id_))){
  tra.by.id=tra[tra$id_==unique(tra$id_)[i],]
  tra.by.id=tra.by.id[order(tra.by.id$t_),]
  sampling.rate[[i]]=summarize_sampling_rate(tra.by.id)
}
sampling.rate
#adjust sampling rate to get regular time steps
tra=tra[order(tra$id_,tra$t_),]
tral=split(tra,tra$id_)
tral=lapply(tral,track_resample,rate=minutes(25),tolerance=minutes(15))
for(i in 1:length(tral)){tral[[i]]$burst_=paste(names(tral)[i],tral[[i]]$burst_)}
tra=do.call("rbind",tral)
head(tra)
#plot points
require(lattice);xyplot(tra$y_~tra$x_|tra$id_)

#1.2) GENERATE RANDOM STEPS FROM A SELECTION-FREE MOVEMENT KERNEL MODEL
#convert from points to steps
ssf=tra %>% steps_by_burst(keep_cols="start")
head(ssf)
#generate random available steps
ssf=ssf %>% random_steps(n = 30)
head(ssf)


##############################################################################
#####PART 2 - ATTACH SPATIAL COVARIATES TO BOTH OBSERVED AND RANDOM STEPS ####
#attach the environmental covariates to both observed and random steps
##load packages
library(raster)
##load map biomas map
map.biomas <- raster("C:/Users/aline/Dropbox/Bandeira/Map/PANTANAL-2015-MAPBIOMAS.tif")
##crop map biomas map
limits <- extent(-56.10,-55.2, -19.50, -19.10)
map.biomas <- crop(map.biomas,limits)
plot(map.biomas)
##project map biomas map
map.biomas <- projectRaster(map.biomas, crs="+proj=utm +zone=21 +south", method='ngb')
##create habitat column 
ssf=ssf %>% extract_covariates(map.biomas,where="end") 
summary(ssf)
#summarize map.biomas column in forests and open 
ssf$map <- ifelse(ssf$PANTANAL.2015.MAPBIOMAS==1|
                           ssf$PANTANAL.2015.MAPBIOMAS==2|
                           ssf$PANTANAL.2015.MAPBIOMAS==3|
                           ssf$PANTANAL.2015.MAPBIOMAS==4|
                           ssf$PANTANAL.2015.MAPBIOMAS==5|
                           ssf$PANTANAL.2015.MAPBIOMAS==9,1,0)

##create a mapS column to represent the habitat used 
ssf$mapS=unlist(lapply(split(ssf,ssf$step_id_),function(x){rep(x$map[1],31)}))
summary(factor(ssf$mapS))

##############################################################################
################################PART 3 - iSSF ANALYSIS #######################

#3.1) GET VARIATES VALUES READY
ssf$log_sl_=log(ssf$sl_+1)
ssf$cos_ta_=cos(ssf$ta_)#ssf$ta_
ssf$map=relevel(factor(ssf$map),"open") 
ssf$mapS=relevel(factor(ssf$mapS),"open")   
ssf$s1=sin(ssf$hour*2*pi/24)
ssf$c1=cos(ssf$hour*2*pi/24)
ssf=time_of_day(ssf,include.crepuscule=F)


#3.2) GET MOVEMENT PARAMETERS READY
sl_distr(ssf)
ta_distr(ssf)

#3.3) RUN MODEL 
basic=fit_issf(case_~
                 #BIASED KERNEL HABITAT
                 map*c1*tpt
               +log_sl_*tpt*tod_end_
               +sl_*tpt*tod_end_
               +cos_ta_*tpt*tod_end_
               +strata(step_id_),data=ssf,model=T)
summary(basic)

```
![Plot 3 - ssf](https://user-images.githubusercontent.com/91481371/165376777-8349495c-59e8-4392-b778-c03242bc548f.jpeg)

**Comment by bsmity13** [comment, marked as answer]
Hi @AlineGiroux ,

Thanks for your question. Based on your model and your example figure, I think I understand what your goal is with this figure. Let me write out a few assumptions and then make a few things more explicit for anyone else that comes across this discussion thread.

I'm assuming that you want the y-axis of the figure to be log-RSS, i.e., the log of Relative Selection Strength (RSS). RSS(x1, x2) for an iSSF is a measure of how many times more likely a step is to end in the habitat x1 vs. the habitat x2 (given the model). For example, when RSS(x1, x2) = 2, your model is predicting that your animal is twice as likely to take a step ending in x1 than in x2, i.e., it selects x1 over x2. If RSS(x1, x2) = 1, your model is predicting that your animal is equally likely to take a step ending in x1 or x2, i.e., it has no preference (no selection). If RSS(x1, x2) = 0.5, then your model is predicting that your animal is half as likely to take a step ending in x1 than in x2 (avoiding x1) -- or equivalently, it is twice as likely to take a step ending in **x2** than it is to take a step ending in **x1** (i.e., 1/0.5 = 2; selecting x2). 

You can see that RSS > 1 is selection, RSS == 1 is no preference, and RSS < 1 is avoidance. RSS is typically more intuitive to understand, but log-RSS is easier to calculate and plot. That's because (1) the model is linear on the log scale; and (2) with RSS, selection can be any value between 1 and ∞, but avoidance is crammed into the space between 1 and 0. Plotting it on the log scale makes the regions of selection and avoidance more equivalent. Anyone interested in details of interpretation using RSS should see [Fieberg et al. 2021](https://doi.org/10.1111/1365-2656.13441) and [Avgar et al. 2017](https://doi.org/10.1002/ece3.3122).

I'm also assuming you want to illustrate the _movement-free habitat selection kernel_ with this figure. It appears to me that your example figure is showing selection for the forest/open covariate (`map`) for different temperatures (`tpt`) at different times of day (`c1`). Your model formula is a three-way interaction (`map * c1 * tpt`), where `c1` is a cosine transformation of hour of the day. So I'm assuming that you want to know how many times more likely your animal is to step into `map = 1` (forest?) than `map = 0` (open?), given a temperature and time of day. So how do we do that?

We have a function called `log_rss()` in `amt` that will calculate log-RSS(x1, x2) from a fitted model for you. You define `x1` and `x2` as `data.frame`s with all of the covariates in your model. Under the hood, these `data.frame`s get passed to `predict()` via the argument `newdata`, so this works almost the same as making predictions from an `lm` or `glm` in `R`. E.g., you have to include all of the variables in the model, any factor variables must have the exact same levels as when the model was fitted, etc. 

The tricky thing in this case is that we enforce `nrow(x2) == 1`. The reason is to prevent a user from making unintended mistakes with R's recycling rules. If you're not sure what I mean, try this in R:

```
x1 <- c(10, 9, 8, 7, 6, 5) 
x2 <- c(1, 2)
x1 - x2
```
As long as `nrow(x2)` is a divisor of `nrow(x1)`, it will _quietly_ recycle `x2` until it is the same length as `x1`. Something similar happens in the log-RSS calculation, so we force `x2` to have exactly 1 row to avoid any possible silent recycling (user) errors. Given that restriction, the best way to calculate log-RSS is via a bunch of separate calculations that you recombine later.

As with any programming problem, there are many ways to go about that, some of which will be easier to understand or more efficient than others. Here's how I would do it. It's unfortunately a bit slow because of the repeated calls to `log_rss()`.

```
library(dplyr)

# Make a base data.frame to create x1 and x2 from:
# Start with all the combinations of hour and tpt
base <- expand.grid(hour = seq(0, 24, length.out = 50),
                     tpt = c(15, 22.5, 30)) %>% 
  # Convert hour to 'c1'
  mutate(c1 = cos(hour * 2*pi/24)) %>% 
  # Add the movement and tod variables 
  #  (note, I'm being very coarse with tod_end_)
  #  (note, it doesn't matter what values you pick for sl_ and ta_)
  mutate(sl_ = 45,
         log_sl_ = log(45),
         cos_ta_ = 1,
         tod_end_ = ifelse(hour < 24 & hour > 11, "day", "night"))

# Remember, factors need to be coded the same in the new data as in 'ssf'
base$tod_end_ <- factor(base$tod_end_, levels = levels(ssf$tod_end_))

# Now we can create x1 and x2, with the only difference being map = 1 in x1
# and map = 0 in x2.
x1 <- base %>% 
  mutate(map = 1)

x2 <- base %>% 
  mutate(map = 0)

# Now we are ready for 'log_rss()', but note we need to go one row at a time
# (and note that x1 and x2 have the same number of rows). E.g., this works:
log_rss(basic, x1[1,], x2[1,], ci = "se")

# The resulting object is a list, and element $df is what you want for plotting.
# It looks like this:

#> $df
#> hour_x1 tpt_x1 c1_x1 sl_x1 log_sl_x1 cos_ta_x1 tod_end_x1 map_x1     log_rss         lwr        upr
#>       0     15     1    45  3.806662         1      night      1 -0.01048721 -0.06022541 0.03925098


# So how do we iterate this? We could use a for loop, we could use lapply, etc.
# I will use lapply here, with a custom function inside.

log_rss_list <- lapply(1:nrow(x1), function(i) {
  # Report status
  cat(i, "of", nrow(x1), "\n")
  # Calculate log-RSS for that row
  xx <- log_rss(basic, x1[i,], x2[i,], ci = "se")
  # Return the element $df
  return(xx$df)
})

# We now have a list with an element for each row. We can combine those rows
# into a single data.frame with 'dplyr::bind_rows()'
res <- bind_rows(log_rss_list)

# This object is ready for plotting. For example:
library(ggplot2)

ggplot(res, aes(x = hour_x1, y = log_rss, color = factor(tpt_x1))) +
  geom_line(size = 1) +
  geom_hline(yintercept = 0, linetype = "dashed") +
  xlab("Hour of Day") +
  ylab("log-RSS") +
  scale_color_discrete(name = "Temperature (°C)") +
  theme_bw()
```
I hope that answers your question. Happy to answer any follow-ups.

**Reply by bsmity13** [reply]
Here's the figure:

![fig](https://user-images.githubusercontent.com/34109166/166841384-9c03400d-119b-42ae-b141-0f41e5fb9d18.png)

**Comment by AlineGiroux** [comment]
Hi @bsmity13, 
Thank you very much. Sorry I was not clearer in my question, but you understood perfectly well. It help me a lot. 
Now I can plot the log of Relative Selection Strength (RSS) throghout the day under some environmental temperature values. 

I also would like to plot the gamma distribution of step lengths and the von mises distribution of turn angles under some environmental temperature values at day and night (please see my example bellow). But I am not sure how can I extract the parameters values of the distributions under different conditions (e.g., period of day and temperature). Can you help with this too, please? 

Thank you very much, I really appreciate your time.

![issf_complete](https://user-images.githubusercontent.com/91481371/167165907-96ecced6-c91a-4a2a-a9b7-1687b5e817d6.jpg)

**Comment by bsmity13** [comment]
Hi @AlineGiroux,

Good question. This is tricky because of your interactions.

As you know (but for anyone else who finds this later), you modeled your tentative steps with a gamma distribution, so you included step length (`sl_`) and the log of step length (`log_sl_`) in your model formula to update the scale and shape parameters (respectively) of the gamma distribution. For anyone unfamiliar with how to do this, a handy reference is [Appendix C of Fieberg et al. 2021](https://conservancy.umn.edu/bitstream/handle/11299/218272/AppC_iSSA_movement.html). If you _didn't_ have an interaction, you would just follow the formulas in the appendix, but you can also use the function `amt::update_sl_distr()` to update the step-length distribution or the function `amt::update_ta_distr()` to update the turn-angle distribution. 

So how do you do this when you have interactions? There are examples of this given in [Appendix B of Fieberg et al. 2021](https://conservancy.umn.edu/bitstream/handle/11299/218272/AppB_SSF_examples.html?sequence=26&isAllowed=y#Movement_and_Habitat_Interactions), but I will expand a bit here.

Repeating your model formula from above:

```
basic=fit_issf(case_~
                 #BIASED KERNEL HABITAT
                 map*c1*tpt
               +log_sl_*tpt*tod_end_
               +sl_*tpt*tod_end_
               +cos_ta_*tpt*tod_end_
               +strata(step_id_),data=ssf,model=T)
```

We see that you have `sl_` in a three-way interaction with temperature (`tpt`) and time-of-day of the end of the step (`tod_end_`). The same goes for `log_sl_` and `cos_ta_`, so the following applies to all 3.

(Quick side note: we tend to model interactions with movement using covariates from the start of the step. But you're free to use `tod_end_` if that matches your biological interpretation better -- maybe your animal's step length depends on what time it will be when it arrives at a destination?)

First, let's note that the `*` in the model formula tells `R` to construct interaction terms with all of the lower order terms, as well, whereas the `:` in the model formula just creates the interaction term. I.e., these two formulas are the same:

```
resp ~ a * b * c
resp ~ a + b + c + a:b + a:c + b:c + a:b:c
```
Each of those covariates gets a slope parameter, so you're actually estimating 7 parameters for that three-way interaction. (Also note that your model is a bit overparameterized; `summary(basic)` shows `NA` for those terms that cannot be estimated).

The trick is to express the parameter of interest (*e.g.*, the beta for `sl_`) as a function of the other covariates in your model (`tpt` and `tod_end_`). I wrote a separate document about this, complete with equations, [here](https://bsmity13.github.io/intxn_fun/). Based on the logic outlined there, when you express the beta for `sl_` as a function of `tpt` and `tod_end_`, you get a different gamma distribution for any possible combination of `tpt` and `tod_end_`. Here's one way to code that:

```
# Grab the coefficients (b for beta)
b <- coef(basic)

# coef for sl_ as a function of tpt and tod_end_
coef_sl_ <- function(b, tpt, tod_end_) {
  
  res <- b[["sl_"]] +
    b[["tpt:sl_"]] * tpt +
    b[["tod_end_night:sl_"]] * (tod_end_ == "night") +
    b[["tpt:tod_end_night:sl_"]] * tpt * (tod_end_ == "night")
  
  return(res)
}

# E.g.,
coef_sl_(b, tpt = 15, tod_end_ = "night")
```

It's basically the same for `log_sl_`, but note you need to get the names of the betas (`b[["name"]]`) right, and sometimes the order switches.

```
# coef for log_sl_ as a function of tpt and tod_end_
coef_log_sl_ <- function(b, tpt, tod_end_) {
  
  res <- b[["log_sl_"]] +
    b[["tpt:log_sl_"]] * tpt +
    b[["log_sl_:tod_end_night"]] * (tod_end_ == "night") +
    b[["tpt:log_sl_:tod_end_night"]] * tpt * (tod_end_ == "night")
  
  return(res)
}

# E.g.,
coef_log_sl_(b, tpt = 15, tod_end_ = "night")
```

Now that you have expressed the betas as functions (and literally as R functions), you can use the function `amt::update_gamma()` to get a single new gamma distribution.

```
# Now pass those to update_gamma() instead of update_sl_distr()
update_gamma(dist = sl_distr(basic),
             beta_sl = coef_sl_(b, tpt = 15, tod_end_ = "night"),
             beta_log_sl = coef_log_sl_(b, tpt = 15, tod_end_ = "night"))
```

To make a nice figure with several combinations, you could do something like this:

```
# Start with the combinations of temperature and time of day
sl_fig <- expand.grid(tpt = c(15, 22.5, 30),
                      tod_end_ = c("day", "night")) %>% 
  # Calculate the betas for each
  mutate(beta_sl = coef_sl_(b, tpt = tpt, tod_end_ = tod_end_),
         beta_log_sl = coef_log_sl_(b, tpt = tpt, tod_end_ = tod_end_))

# For each row, get the updated shape and scale for the selection-free step-length
# distribution (this function is vectorized)
sf_dist <- update_gamma(sl_distr(basic),
             beta_sl = sl_fig$beta_sl,
             beta_log_sl = sl_fig$beta_log_sl)

sl_fig$shape <- sf_dist$params$shape
sl_fig$scale <- sf_dist$params$scale

# Now get likelihood from gamma distribution for a bunch of possible steps
sl <- seq(0, 200, length.out = 100)
sl_fig <- merge(sl_fig, y = data.frame(sl = sl)) %>% 
  mutate(like = dgamma(sl, shape = shape, scale = scale))

# Plot
ggplot(sl_fig, aes(x = sl, y = like, color = factor(tpt))) +
  facet_wrap(~ tod_end_) +
  geom_line(size = 0.9) +
  xlab("Step Length (m)") +
  ylab("Likelihood") +
  scale_color_viridis_d(name = "Temperature (°C)") +
  theme_bw()
```

Which results in this figure:

![fig2](https://user-images.githubusercontent.com/34109166/167225977-88fdd111-edc0-4f2e-b8f9-9bcec73e168c.png)

I hope that is clear. This can be a confusing thing to get your head wrapped around the first time, so this discussion should be useful for a lot of people. Thanks for asking.
